### PR TITLE
[Opmondev]:171 Opendata statistics aggragator script and API endpoint

### PR DESCRIPTION
**Pull request by wisecrow**
1. Create a script to count requests in MongoDB `raw_messages` and store single document with counts in MongoDB collection `metrics_statistics`

    ` xroad-metrics-statistics ` 

Example document:
```
{
    "_id" : ObjectId("64afd4d71dc8487dba2b3d45"),
    "currentMonthRequestCount" : 3742,
    "currentYearRequestCount" : 4648,
    "previousMonthRequestCount" : 872,
    "previousYearRequestCount" : 0,
    "todayRequestCount" : 0,
    "totalRequestCount" : 4648,
    "updateTime" : 1689244887778.66
}
```

2. Pass optional parameter `output_only` to output counts only without storing it to MongoDB:
` xroad-metrics-statistics --output_only`

Example output:
```
2023-07-14 14:00:14,159 - INFO - Metrics statistics data: starting aggregate
2023-07-14 14:00:14,206 - INFO - Metrics statistical data:

{ 'current_month_request_count': 3720,
  'current_year_request_count': 4648,
  'previous_month_request_count': 894,
  'previous_year_request_count': 0,
  'today_request_count': 0,
  'total_request_count': 4648}
```

 3. Create API endpoint `api/statistics` to get stored document and return as json.

```
curl --location 'http://localhost:8081/api/statistics'
{"current_month_request_count": 3720, "current_year_request_count": 4648, "previous_month_request_count": 894, "previous_year_request_count": 0, "today_request_count": 0, "total_request_count": 4648, "update_time": 1689343228154.687}
```

### CODE REVIEW

File: opendata_module/metrics_statistics/mongodb_manager.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
+""" MongoDB Manager - Opendata Module
+"""
```
Comment (justinask7) [inline opendata_module/metrics_statistics/mongodb_manager.py:2]:
```suggestion
""" MongoDB Manager - Opendata Module"""
```

File: opendata_module/metrics_statistics/mongodb_manager.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+
+import logging
+import urllib.parse
+from datetime import datetime, timedelta
+from typing import Any, Dict, List, Mapping, Optional, TypedDict
```
Comment (justinask7) [inline opendata_module/metrics_statistics/mongodb_manager.py:29]: Optional: imports doesn't seem in a correct order 🤔

File: opendata_module/metrics_statistics/mongodb_manager.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        return metrics_ts(self.to_dt)
+
+
+class RequestCountDateTimeRange(TypedDict):
+    total_request_count: None
+    previous_year_request_count: DateTimeRange
+    current_year_request_count: DateTimeRange
+    previous_month_request_count: DateTimeRange
+    current_month_request_count: DateTimeRange
+    today_request_count: DateTimeRange
```
Comment (justinask7) [inline opendata_module/metrics_statistics/mongodb_manager.py:77]: Optional: it's quite convenient to have typeddicts at the top of file and only then have classes

File: opendata_module/metrics_statistics/mongodb_manager.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    @staticmethod
+    def get_today_range() -> DateTimeRange:
+        now = datetime.now()
+        today_start = datetime(now.year, now.month, now.day, 0, 0, 0)
+        return DateTimeRange(today_start)
+
+
+class DatabaseManager:
+
+    def __init__(self, mongo_settings: dict,
```
Comment (justinask7) [inline opendata_module/metrics_statistics/mongodb_manager.py:124]: Hmm, shouldn't it be `Dict`?

File: opendata_module/metrics_statistics/settings_parser.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+
+import errno
+from os import listdir
+from os.path import isfile, join
+import re
+from copy import deepcopy
```
Comment (justinask7) [inline opendata_module/metrics_statistics/settings_parser.py:27]: Small point - import order

File: opendata_module/metrics_statistics/settings_parser.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    Class to hold X-Road Metrics user settings.
+
+    Can parse settings from a YAML file.
+    Settings file is searched from the current working directory and /etc/xroad-metrics/opendata/.
+    Settings file must have extension .yaml or .yml.
+    If profile argument is set, settings are fetched from settings_{profile}.yaml.
+    If no profile is defined, settings are fetched from settings.yaml.
+    """
+
+    def __init__(self, profile=None):
```
Comment (justinask7) [inline opendata_module/metrics_statistics/settings_parser.py:43]: types would be nice in this whole class

File: opendata_module/opmon_opendata/mongodb_manager.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+""" MongoDB Manager - Opendata Module
+"""
```
Comment (justinask7) [inline opendata_module/opmon_opendata/mongodb_manager.py:2]:
```suggestion
""" MongoDB Manager - Opendata Module"""
```

File: opendata_module/opmon_opendata/mongodb_manager.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+import logging
+import urllib.parse
+from datetime import datetime, timedelta
+from typing import Mapping, Optional, TypedDict
+
+from pymongo import MongoClient
+
+
+class StatisticsDataNotFound(Exception):
```
Comment (justinask7) [inline opendata_module/opmon_opendata/mongodb_manager.py:34]:
```suggestion
class StatisticsDataNotFoundError(Exception):
```

**Review by justinask7 — APPROVED** [review]

File: opendata_module/metrics_statistics/mongodb_manager.py (hunk 2 above)
Comment (wisecrow) [reply opendata_module/metrics_statistics/mongodb_manager.py:29]: Maybe I missing something, but imports looks sorted in alphabetical order :)

Comment (wisecrow) [reply opendata_module/metrics_statistics/mongodb_manager.py]: Fixed, thanks!

File: opendata_module/metrics_statistics/mongodb_manager.py (hunk 4 above)
Comment (wisecrow) [reply opendata_module/metrics_statistics/mongodb_manager.py]: Add more strict type hint with `Dict`

**Pull request merged**